lib/Sema/CSSolver.cpp

```diff
@@ -2167,40 +2162,51 @@ bool ConstraintSystem::solveRec(SmallVectorImpl<Solution> &solutions,
   }
 
   // Sort the constraints into buckets based on component number.
-  std::unique_ptr<ConstraintList[]> constraintBuckets(
-                                      new ConstraintList[numComponents]);
+  std::unique_ptr<ConstraintBucket[]> buckets(
+      new ConstraintBucket[numComponents]);
+
   while (!InactiveConstraints.empty()) {
     auto *constraint = &InactiveConstraints.front();
     InactiveConstraints.pop_front();
-    constraintBuckets[constraintComponent[constraint]].push_back(constraint);
+    buckets[constraintComponent[constraint]].record(constraint);
   }
 
+  // Create component ordering based on the information associated
+  // with constraints in each bucket - e.g. number of disjunctions.
+  std::vector<unsigned> componentOrdering;
+  // First seed the ordering with original indexes of the components.
+  for (unsigned component = 0; component != numComponents; ++component)
+    componentOrdering.push_back(component);
+  // Now sort the components in the ordering based on the scores of the buckets.
+  std::sort(
```

Comment:
I've separated ordered set from the original one because `components` is a vector and some of the code down the line relies on the original indexes the way they are produced by constraint graph e.g. associating constraints into their buckets and orphan handling logic. I think that should be fixed down the road but I just didn't want to tackle it right away...